interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,367 @@
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const customDeviceSummary = `provides access to specific devices via the gadget snap`
+
+const customDeviceBaseDeclarationPlugs = `
+  custom-device:
+    allow-installation: true
+    allow-connection:
+      slot-attributes:
+        custom-device: $PLUG(custom-device)
+    allow-auto-connection: false
+`
+
+const customDeviceBaseDeclarationSlots = `
+  custom-device:
+    allow-installation:
+      slot-snap-type:
+        - gadget
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+var (
+	// A cryptic, uninformative error message that we use only on impossible code paths
+	customDeviceInternalError = errors.New(`custom-device interface internal error`)
+
+	// Validating regexp for filesystem paths
+	customDevicePathRegexp = regexp.MustCompile(`^/[^"@]*$`)
+
+	// Validating regexp for udev device names.
+	// We forbid:
+	// - `|`: it's valid for udev, but more work for us
+	// - `{}`: have a special meaning for AppArmor
+	// - `"`: it's just dangerous (both for udev and AppArmor)
+	// - `\`: also dangerous
+	customDeviceUDevDeviceRegexp = regexp.MustCompile(`^/dev/[^"|{}\\]+$`)
+
+	// Validating regexp for udev tag values (all but kernel devices)
+	customDeviceUDevValueRegexp = regexp.MustCompile(`^[^"{}\\]+$`)
+)
+
+// customDeviceInterface allows sharing customDevice between snaps
+type customDeviceInterface struct{}
+
+func (iface *customDeviceInterface) validateFilePath(path string) error {
+	if !customDevicePathRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device path must start with / and cannot contain special characters: %q`, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`custom-device path is not clean: %q`, path)
+	}
+
+	if _, err := utils.NewPathPattern(path); err != nil {
+		return fmt.Errorf(`custom-device path cannot be used: %v`, err)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateDevice(path string) error {
+	// The device must satisfy udev's device name rules and generic path rules
```

Comment:
No, they are not, but I suggest we support those rules who are. The documentation is [here](https://www.freedesktop.org/software/systemd/man/udev.html#*), and the [source code](https://github.com/systemd/systemd/blob/3e7bf8535f991e6a8e4104e7b6303b783b372d21/src/udev/udev-rules.c#L1372) shows that udev is just calling `fnmatch()` with `0` flag. So:
- `?`: works the same as in AppArmor, but can also match a `/`. I doubt that one would write a rule where the `/` can match a slash, so I think we are not being very restrictive if we make it work exactly like AppArmor's;
- `*`: is equivalent to AppArmor's `**`; my suggestion therefore is to document that the `*` (both in the `udev-tagging` and the `devices` attributes) also matches `/`, so we basically transform it into a `**` when we translate the `devices` attribute into AppArmor rules;
- `[]`: equivalent, but I now noticed in the [glob man page](https://man7.org/linux/man-pages/man7/glob.7.html) that `fnmatch()` (and therefore udev) uses `!` as the negation character, and not `^`; therefore, to keep thing simpler, I would forbid both `^` and `!`; I don't think that negation is common in udev rules and device names.
- `|`: I decided not to support it at least for this first iteration, to keep things simpler, but I don't think it would be hard to support later, if required. It will end up generating separate AppArmor rules, that's all.

Reply:
> *: is equivalent to AppArmor's **; my suggestion therefore is to document that the * (both in the udev-tagging and the devices attributes) also matches /, so we basically transform it into a ** when we translate the devices attribute into AppArmor rules;

this difference is probably where the plan breaks down. OTOH we might just accept the difference and still translate * to *, udev would match more but in the end there's not a lot of nesting going on in /dev

Reply:
OK. I already changed the branch to add the `*` -> `**` translation, let me know if you prefer it to be reverted (it's 1 line anyway).

Reply:
yes, my preference is to "\*" -> "\*"